### PR TITLE
Allow restricting possible runtime references

### PR DESCRIPTION
**Pull request by ncfavier**
Fixes https://github.com/NixOS/nix/issues/5633

Adds a new derivation attribute `referenceablePaths` which restricts the list of paths that are searched in build outputs in order to establish runtime references.

Useful when creating filesystem images containing their own embedded Nix store: setting `referenceablePaths = [];` makes them self-contained blobs of data with no runtime dependencies.

To do:
- [ ] allow specifying output names just like in `allowedReferences`. Not clear if this is useful, but why not. We can intersect the list of output names with the keys of `scratchOutputs` when adding those to the set.
- [ ] should we have `unreferenceablePaths` as well? (I don't think so)

### CODE REVIEW

**Comment by ncfavier** [comment]
I think CI is failing because the daemon doesn't have the patch? Checks pass locally

File: tests/check-refs.sh [hunk 1]
```diff
@@ -40,3 +40,7 @@ nix-build -o $RESULT check-refs.nix -A test7
 
 # test10 should succeed (no disallowed references).
 nix-build -o $RESULT check-refs.nix -A test10
+
+# test11 should succeed (no referenceable paths)
+test11=$(nix-build -o $RESULT check-refs.nix -A test11)
+! nix-store -q --references "$test11" | grep -q .
```
Comment (thufschmitt) [inline tests/check-refs.sh:46]:
```suggestion
[[ -z $(nix-store -q --references "$test11") ]]
```

might be a bit easier to understand

File: src/libstore/build/local-derivation-goal.cc [hunk 2]
```diff
@@ -2138,12 +2138,29 @@ DrvOutputs LocalDerivationGoal::registerOutputs()
...
+            if (worker.store.isStorePath(i))
+                spec.insert(worker.store.parseStorePath(i));
+            else
+                throw BuildError("derivation contains an illegal possible reference '%s'", i);
+        }
+
+        StorePathSet tmp;
+        std::set_intersection(referenceablePaths.begin(), referenceablePaths.end(),
+                              spec.begin(), spec.end(),
+                              std::inserter(tmp, tmp.begin()));
```
Comment (thufschmitt) [inline src/libstore/build/local-derivation-goal.cc:2160]: I think it would be useful to also log (`notice` I would say) the references that we erase here

File: tests/check-refs.sh (hunk 1 above)
Comment (thufschmitt) [inline tests/check-refs.sh:46]: You probably want to use [isDaemonNewer](https://github.com/NixOS/nix/blob/cacfc32ee22bacb6400796e47fd4dee41299b66a/tests/common.sh.in#L133-L138) to only test this if the daemon is recent-enough. Maybe something like:

```suggestion
if isDaemonNewer 2.12pre20221001; then
  # test11 should succeed (no referenceable paths)
  test11=$(nix-build -o $RESULT check-refs.nix -A test11)
  ! nix-store -q --references "$test11" | grep -q .
fi
```

File: tests/check-refs.nix [hunk 3]
```diff
@@ -8,44 +8,49 @@ rec {
     name = "check-refs-" + toString nr;
   } // args);
 
-  src = builtins.toFile "aux-ref" "bla bla";
+  src = mkDerivation {
+    name = "aux-ref";
+    buildCommand = ''
+      echo "bla bla" > "$out"
+    '';
+  };
```
Comment (thufschmitt) [inline tests/check-refs.nix:16]: Is this actually needed?

File: doc/manual/src/language/advanced-attributes.md [hunk 4]
```diff
@@ -2,6 +2,26 @@
...
 
+  - [`referenceablePaths`]{#adv-attr-referenceablePaths}\
+    The optional attribute `referenceablePaths` restricts the set of
+    store paths whose cryptographic hash Nix looks for in the build
+    outputs in order to establish runtime dependencies. By default,
+    this set includes build inputs, output paths, and paths built by
+    recursive Nix invocations.
+    If this attribute is specified, then this set is restricted (in the
+    sense of set intersection) to the given list of paths.
+    For example,
```
Comment (thufschmitt) [inline doc/manual/src/language/advanced-attributes.md:13]: I find this doc a bit too low-level. Maybe something around the lines of

```suggestion
  - [`referenceablePaths`]{#adv-attr-referenceablePaths}\
    The optional attribute `referenceablePaths` restricts the set of
    store paths that Nix will consider when scanning for the runtime
    dependencies.
    If the attribute is unset, then this set will consist of all the build
    inputs, output paths, and paths built by recursive Nix invocations
    
    For example,
```

(But it's great to have some doc without even having to ask for it! 😃 )

File: tests/check-refs.sh [hunk 5]
```diff
@@ -40,3 +40,7 @@ nix-build -o $RESULT check-refs.nix -A test7
 
 # test10 should succeed (no disallowed references).
 nix-build -o $RESULT check-refs.nix -A test10
+
```
Comment (thufschmitt) [inline tests/check-refs.sh:43]: Could you also add a test with a non-empty `referenceablePaths` (and making sure that a path in it will be properly considered)?

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Thanks, that's a much needed thing, glad to see a PR for that!

Left a few comments/suggestions, but overall 👍

File: tests/check-refs.nix (hunk 3 above)
Comment (ncfavier) [reply tests/check-refs.nix]: I guess not

File: src/libstore/build/local-derivation-goal.cc (hunk 2 above)
Comment (ncfavier) [reply src/libstore/build/local-derivation-goal.cc:2160]: I'm a bit skeptical: I don't think it adds any information that the user doesn't already know, and I think there are usually too many paths in `referenceablePaths` for this to be useful and not spam.

**Comment by ncfavier** [comment]
I renamed the attribute to `possibleReferences`.

**Comment by edolstra** [comment]
I'm not in favor of this. It breaks a fundamental property of Nix, namely the closure invariant, i.e. every reference of a valid path is also present and valid, and thus there are no "gaps" in closures. It decouples the references stored in the Nix DB from the actual references in the filesystem.

I realize that the main use case here is filesystem images, where it is desirable and reasonable to ignore references, but this PR also opens the possibility of *optional* runtime dependencies and is sure to be (ab)used to for that. E.g. if I have a package like `digikam` that has an expensive reference to `marble` (which is very big), I could mark `marble` as an ignored dependency. So `digikam` will be able to use `marble` if it happens to be present. But that means that closures no longer behave in a predicable way; their behaviour would depend on other paths that may be present in the store.

A minor issue is that it's currently at least theoretically possible to recover the references graph (i.e. the contents of the Nix database) simply by scanning all store paths for references to all other store paths. With this PR, that property no longer holds since you need the `possibleReferences` attribute of the derivations that built the store paths. So it wouldn't be possible anymore to have a `nix rebuild-db` command.

**Comment by ncfavier** [comment]
Regarding the first point, Nix does not have that property. I can easily make a program that effectively references a store path in a way that is not detected by the scanner, simply by mangling the hash. We are using a *heuristic* to detect references, which is sometimes faulty, and this PR allows overriding the heuristic when it is faulty. This does not change any fundamental property of Nix. If you fear that it would be abused, we can always add a big red warning in the documentation.

I agree that the second point is more of a problem. How about, as an alternative, storing the list of possible references in a designated path under the output tree, say `/nix-support/possible-references`? Pros: reference scanning is deterministic. Obvious cons: only works for derivations producing directories.

**Comment by thufschmitt** [comment]
I agree that this isn't really breaking the closure invariant since we can already `rot13` (or any reversible transformation) twice the output to get rid of the dependencies. It's just making this less hacky for the cases where it's needed.

@edolstra Is the `digikam` example something you've seen asked for in practice? It feels to me that if I really wanted this kind of optional dependencies and couldn't afford a proper wrapper I'd rather rely on an env variable pointing to some path outside of the store (like what NixOS does with the opengl drivers or just any script relying on `$PATH`) rather than hardcoding a possibly existing path

If that's the case, would it be reasonable to just make it possible to ignore *all* the references? That should solve the most common use-case while not allowing this kind of optional dependencies pattern

**Comment by edolstra** [comment]
Of course you can hide dependencies, just as you can hide pointers to a conservative garbage collector by negating the bits in the pointer, but then it's kind of clear you're doing something naughty. This would make dangling references a supported feature and that's a pretty fundamental change.

At the very least, this should be an experimental feature.

**Comment by ncfavier** [comment]
I'm not sure an experimental feature really makes sense for this.

If the intent is to make sure it's not "too easy" to (ab)use, how about giving it a scary name like `unsafeRestrictReferences` and not including it in the documentation, like `unsafeDiscardStringContext` (which can also be used for optional dependencies)?

**Comment by Ericson2314** [comment]
> A minor issue is that it's currently at least theoretically possible to recover the references graph (i.e. the contents of the Nix database) simply by scanning all store paths for references to all other store paths. 

That's just not true.

1. We would need an exception for drv files. The output paths in drv files do *not* count as references.

2. Anyone is free to manually type a non-present nix store path in a build, e.g. `/nix/store/himynameisjohn-foobar` and arrange for it to be in an output. References of outputs are always a *subset* of drv inputs. Since things string merely looks like a store path but isn't a drv intput it won't count. But if we were to scan for references after the fact as you purpose, we would have no reason not to count it!
   
    With input-addressed paths, scanning for references to rebuild the DB can in fact find spurious cycles this way!

Bottom line is, the references are *not* mere metadata, or a mere cache of scanning. References are data proper, as much part of the store object as its file system objects are. This is the only correct way to account for how Nix actually behaves.

**Comment by Ericson2314** [comment]
Experimental feature still makes sense, however. By default things should start as experimental features this is good.

**Comment by ncfavier** [comment]
I've renamed the attribute to `__visibleReferences`, hidden it under the experimental feature `invisible-references` and documented it as experimental.

I'm not sure what else needs to be done, e.g. should this interact with `requiredSystemFeatures` somehow?

**Comment by thufschmitt** [comment]
> I'm not sure what else needs to be done, e.g. should this interact with requiredSystemFeatures somehow?

That could be good (so that you don't dispatch a build to a builder that doesn't have the xp feature), but it's not a blocker imho. And maybe not really needed at all since I hope that this won't stay experimental for too long (ideally it could ship as Xp for the next release and then stable in the following one)

**Review by thufschmitt — APPROVED** [review]

**Comment by ncfavier** [comment]
Not sure how to get the tests to work, `enableFeatures invisible-references` doesn't work either.

**Comment by thufschmitt** [comment]
> Not sure how to get the test daemon to pick up the experimental feature, enableFeatures invisible-references doesn't work either.

If you haven't done that, I think you need to call `restartDaemon` after that to force the daemon to take the new config into account

**Comment by ncfavier** [comment]
Is anything else blocking this?

**Comment by Ericson2314** [comment]
@ncfavier @edolstra will need to be convinced, or delegate the decision, or similar. This is in backlog while the Nix team is getting going.

**Comment by roberth** [comment]
Here's another case.
 - https://github.com/NixOS/nixpkgs/pull/194345
 
It'd be much easier to test as well, if it wouldn't pull in all the contents as transitive dependencies of the test.

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting 2022-11-28:

<details><summary>Complete discussion</summary>

- @thufschmitt: @edolstra argues [this breaks the closure invariant](https://github.com/NixOS/nix/pull/7087#issuecomment-1257730655)
    - this does not really hold, because it's an invariant on the build inputs and one can already trick the reference scanner
    - @tomberek: the question is how difficult it is in practice. the tricks make evident that it's something unsafe to do.

- @ericson2314: Underlying philosophical question is: are the references of the store object *normative* or are they merely a *cache* of scanning: the underlying data is normative.
    - @thufschmitt: one could consider it the cache. the scanner already takes the set of build inputs, that's how we restrict it. so `__visibleReferences` is just additional information on the restriction

  - Existing ways to "get around" regular scanning:
    - Just hard-code a non-input store path to the build
      - Literally type `/nix/store/asdfasdfasdf-foo`
      - Do `${builtins.unsafeDiscardStringContext storePathExpr}`
    - Compress/encrypt referenes and then uncompress/unencrypt ones.
      - @roberth: tried that and Nix seemed to scan all of the build closure, not just the closure of outputs of inputs 
        - therefore that workaround doesn't work

- @roberth: the use cases listed are all about self-contained blobs that may contain store paths.
    - we could simplify the feature by adding a type of derivation that has no references.
    - @toberek: that could be implemented by skipping the scan phase

    - could have a flag `disableScanning` or `selfContained`
- @Ericson2314: would like to have it consistenlty: right now adding files to the store will never have references, but creating store objects through derivations will always have references. this should be uniform, and both should have a flag to either scan or not scan for references.
    - this is what currenlty happens:
        ```
        nix-repl> builtins.toFile "asdf" "${pkgs.hello}"
        error: in 'toFile': the file named 'asdf' must not contain a reference to a derivation but contains (!out!/nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv)
        ```
    - @fricklerhandwerk: this will IFD if the store path doesn't exist

    - this will not IFD - and works! Added sources already can have references:
        ```
        nix-repl> builtins.toFile "asdf" "${builtins.toFile "asdf" "bar"}"
        "/nix/store/vx1w2qd1l6imzqz4r6iqm11nkab3vhi5-asdf"
        ```
        ```
         $ nix --extra-experimental-features nix-command path-info /nix/store/vx1w2qd1l6imzqz4r6iqm11nkab3vhi5-asdf --json | jq
        [
          {
            "path": "/nix/store/vx1w2qd1l6imzqz4r6iqm11nkab3vhi5-asdf",
            "narHash": "sha256-DrUdYd4+l2+oIWKUTq+YBwQclIiQRo8+7AISwvZiKuo=",
            "narSize": 160,
            "references": [
              "/nix/store/gjigc1084dfknf8z1yv4i8sjgpjai7ih-asdf" # SEE HERE
            ],
            "ca": "text:sha256:0rv5brchrzczvqv5y2gnzv2a818wz8lahb3nzgb9axra766lh4ws",
            "registrationTime": 1669641171
          }
        ]
        ```

- @roberth: tangent: it may be useful to have a builtin for producing directories
    - https://github.com/NixOS/nix/issues/3497
    - in the context of the matrix: evaluation/build time, references/no references
        - this is the related problem of symmetry
- @fricklerhandwerk: should it be a "magic" (`__<name>`) attribute?
    - @thufschmitt: all attributes that change the behavior should be magic, because all others are passed as inputs to the builder
   
</details>

Decision: @ncfavier Please make the attribute a boolean, not a list. Instead of filtering down to a set of references, declare in the derivation that its build output has no references.

We think we can skip scanning entirely in this case, but we are not sure.
Implementation strategy up to further discussion.

**Comment by ncfavier** [comment]
Done: `__visibleReferences` → `__selfContained`, `invisible-references` → `self-contained-derivations`.

**Comment by roberth** [comment]
The term "self-contained derivation" doesn't feel right. Actually self-contained _derivation_ is wrong, because it's a property of an _output_.
Of course we don't _need_ this feature to be this granular, but doing it at the derivation level instead of the output level _is_ a wart, and it shows in the name. :(
How do we represent properties of an output nicely in the derivation format?

**Comment by fricklerhandwerk** [comment]
@roberth agreed, naming is important and we should be careful not to increase the level of confusion with misleading names.

Random idea: move related properties into an extra attribute `__output`.

**Comment by roberth** [comment]
I wish it was just naming. The right name suggests a different, more flexible implementation.

> `__output`

`__outputs.out.selfContained = true`?

Will this increase implementation complexity? If so, is that because it resolves tech debt?

My apologies if my suggestions seem arbitrary; first simplifying, and now adding different complexity (maybe)? I think we were unaware of the design space, and we're heading in the right direction.

**Comment by ncfavier** [comment]
If we're making this per-output, I might as well bring back the `echo true > $out/nix-support/self-contained` idea. This makes the property explicit, possibly surviving a database rebuild (I haven't kept track of whether this is a concern or not in the end), at the cost of not supporting single-file outputs.

Kind of annoying to implement though. I think the `__outputs.out.selfContained` thing is fine in terms of complexity.

**Comment by Ericson2314** [comment]
We could also restrict to a single output.

**Comment by fricklerhandwerk** [comment]
@roberth yes.

@ncfavier writing a dedicated file would preclude having a single file as the resulting store object, right? Do database rebuilds happen often in practice to care at this point?

Also, it seems fine to get the interface right first, the underlying implementation can still change invisibly.

**Comment by roberth** [comment]
> We could also restrict to a single output.

Doing this is arbitrary and doesn't fix the problem of the inaccurate name.

**Comment by roberth** [comment]
> @ncfavier writing a dedicated file would preclude having a single file as the resulting store object, right? 

It would, but in my opinion single-file store paths are to be avoided anyway. I know of three reasons to use it, neither of which is good.
 - Convenience: being able to do `${writeScript "foo" ...} --letsgo`. This is unnecessary since we could set something along the lines of `passthru.outPath = drv.outPath`. We don't do this today because we conflate derivations and packages, which of course we shouldn't.
 - Laziness: not having to do a `mkdir` first.
 - Legacy: once you go with a single file in a public interface, the only way out is a breaking change

> Do database rebuilds happen often in practice to care at this point?

I don't think a database rebuild is a good solution. The store is meant to be derived state. If derived state is corrupted, you regenerate it, instead of making educated guesses based on some underused code path. Presumably your system is in a good enough state to boot, and then you can reset the database to empty and perform all the builds that are required for the system (usually just nixos, but perhaps home-manager or something else), and then you gc to remove the broken store paths. This way things are pretty much guaranteed to go back to normal.

@edolstra do you know of a situation where this isn't a viable strategy?

**Comment by Ericson2314** [comment]
I could be convinced, but I don't really feel great about `nix-support/self-contained`. To inspect the output, as opposed to the input, to decide how to interpret (hash in this case) the output doesn't feel right.

**Comment by ncfavier** [comment]
Actually the `__outputs.out.selfContained` solution is also annoying because it requires also setting `__structuredAttrs = true;` and the nixpkgs stdenv isn't happy with that. Maybe `__selfContained = [ "out" ];` instead?

This leaves the interface open to making it even more granular in the future, e.g. `__selfContained = [ (builtins.placeholder "out" + "/nixos.iso") ];`, if there's ever a need for that.

**Comment by ncfavier** [comment]
...on the other hand, the `(dis|)allowed(References|Requisites)` attributes already implicitly mean "for all outputs of this derivation", I'm not sure self-contained needs to be different. If the problem is the name of the experimental feature we can just call it `self-contained`.

Also, should the attribute actually start with `__`? I can't really make out a rule for when that is used.

**Comment by fricklerhandwerk** [comment]
> Also, should the attribute actually start with `__`? I can't really make out a rule for when that is used.

The rule is that the value of ever regular attribute name is set as an environment variable for the builder. Therefore parameters controlling behavior should be prefixed with `__`. See at the end of the details section in the [Team discussion comment](https://github.com/NixOS/nix/pull/7087#issuecomment-1337017301).

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting 2023-01-02:

- @edolstra: this is a "dangerous" feature, because the output may actually have references that Nix is not aware of
    - also, if `__structuredAttrs` is enabled, there is an output attribute set mapping a bunch of checks. this is where this feature should hook into
        - should look like this: `<output>.unsafeDiscardReferences = true;`
    - @roberth: Nixpkgs recently made `stdenv` support `__structuredAttrs` to a reasonable degree
    - @fricklerhandwerk: structured attributes are not documented anywhere
        - @edolstra: created an [issue](https://github.com/NixOS/nix/issues/7537)

**Comment by ncfavier** [comment]
> * also, if `__structuredAttrs` is enabled, there is an output attribute set mapping a bunch of checks. this is where this feature should hook into

Done.

I'm not sure it should really go in `outputChecks`, since this feature comes in long before `checkOutputs` is called. Maybe just `unsafeDiscardReferences.out = true`?

**Comment by roberth** [comment]
> I'm not sure it should really go in `outputChecks`,

Another reason is that it's not a check in the sense of `a -> Bool` or `void check(a) throw;`, but rather a transformation.

**Comment by Ericson2314** [comment]
This stalled out, are any more steps needed to be taken, or is it good now?

**Review by thufschmitt — APPROVED** [review]

**Comment by ncfavier** [comment]
- https://github.com/NixOS/nix/pull/7714

**Comment by 7c6f434c** [comment]
Well, this as experimental will not really get us data on the drawbacks of this compared to the safer «declare an output self-contained», as the hiding will need to be very explicitly asked by the user.

Hopefully this will stay experimental…

**Comment by roberth** [comment]
@7c6f434c makes a good point.

Actually having this as experimental makes the feature much worse because any mistakenly unconfigured builder will ignore the attribute and produce an output with a ton of dependencies that will propagate through caches etc. Users won't know why it happened and then think that this feature has a bug.

The real problem is _forward_ compatibility. I don't think we have a way for a derivation to request that it's built by a Nix that has specific support for something, do we? `requiredSystemFeatures` could be that, but I don't know if we want to use it for this purpose, as any new such feature requires remote builder reconfiguration and such.

**Comment by ncfavier** [comment]
> will ignore the attribute

It will [fail](https://github.com/NixOS/nix/pull/7087/files#diff-08a5e40e2243b11f7d074881064ea221e03448cce5bdf0faee0c2fe0f197571dR2329), actually.

I think ignoring the attribute would actually be preferable. People who know about the experimental feature can try it out. Others get the status quo ("ton of dependencies").

**Comment by 7c6f434c** [comment]
> I don't know what we're afraid of or why we should make life so hard for everyone including ourselves (Nix team) who need to learn about this feature's use. Unknown unknowns, really?

Well, it violates a pretty fundamental assumption that everyone considers a foundational principle of Nix. It definitely can be abused badly… Is it that hard to overrideDerivation its use into any image-generating expression, though (provided you have it enabled)?

**Comment by ncfavier** [comment]
It's not hard, provided the image builder supports `__structuredAttrs`: https://github.com/NixOS/nixpkgs/pull/214373

With the above PR I'm able to build a reference-free ISO with `config.system.build.isoImage.overrideAttrs (_: { unsafeDiscardReferences.out = true; })`.

**Comment by roberth** [comment]
> Well, it violates a pretty fundamental assumption that everyone considers a foundational principle of Nix.

So does `unsafeDiscardStringContext`, and yet I have never encountered anyone who has a problem with it.

> It definitely can be abused badly…

Abused sure, anything can be abused, but badly? In a way where the responsibility falls on Nix?

I suggest we remove the experimental flag soon, and if anyone wants to exercise their caution, work on a forward compatibility attribute instead; a _known_ problem.

**Comment by 7c6f434c** [comment]
> Abused sure, anything can be abused, but badly? In a way where the responsibility falls on Nix?

«I have done a GC and now a package I have installed has reduced functionality, isn't this supposed to be impossible» is quite likely to be blamed on Nix.

**Comment by roberth** [comment]
Sure. I'd rather have a useful tool than a safe tool. Let's agree to disagree.

**Comment by 7c6f434c** [comment]
It's not merely about a _safe_ tool, it is about whether a tool gives you guarantees you can quickly understand and _reason about_. Reducing unpredictable interactions is more or less the whole point of Nix, so exceptions to it that can be hidden in an expression reduce the value.

As a useful tool I would prefer a trusted user dropping all the references of a specific path, as this is a clearly informed local choice, and the user could just do this anyway via pushing to a local cache and dropping the references (or prefetching as FOD, or whatever they prefer)

**Comment by ncfavier** [comment]
> «I have done a GC and now a package I have installed has reduced functionality, isn't this supposed to be impossible» is quite likely to be blamed on Nix.

How often do you see people complaining that Nix is not pure enough, or not safe enough? How often do you see people complaining about Nix being too resource-hungry or wasteful? I certainly see the latter way more often.

> exceptions to it that can be hidden in an expression reduce the value.

No they don't. GHC wouldn't exist without `unsafePerformIO`.

Using a **heuristic** for determining references and not providing an escape hatch for false positives in the heuristic is nonsensical.

> a trusted user dropping all the references of a specific path

The burden shouldn't fall on the user though (no matter how trusted or high-up-the-supply-chain). As a derivation author, I know that the hashes appearing in `out` aren't actual references, so I make the informed choice to use `unsafeDiscardReferences`.

**Comment by 7c6f434c** [comment]
> How often do you see people complaining that Nix is not pure enough, or not safe enough? 

I personally? A lot of time. Like a significant part of discussion about flakes is specifically about what is the proper make to get things more pure. I do also see discussions of how to make sandboxing safer.

I do see complaints that are basically «deduplication needs to be improved a lot», which seems to be primary motivation for Content-Addressed work.

Of model-breaking complaints, I mostly see «how do I make one store path writeable and seeded by some nix-build provided contents». Somehow, the idea of actually making it an option doesn't seem to get traction. Some of the issues can be resolved with carefully constructed FHS user envs, but not all of then.

**Comment by RaitoBezarius** [comment]
Hello, we are multiple projects who are interested into that feature:

- `make-disk-image` which I am code owner (shaving ~500MB for a NixOS test disk image) ;
- @alyssais's Spectrum project ;
- nixos-generator's https://github.com/nix-community/nixos-generators/issues/231 project

Unfortunately, the way this feature is designed makes it very hard to let user who want to opt-in into this unsafeness get the benefits.

For example, `unsafeDiscardReferences.out = true;` will eval but not build if you don't have the experimental features, this prevent graceful degradation to the usual behavior, which would be nice.

Furthermore, if we want to have it in the Hydra buildfarm for our current disk images, this would make it very hard because of the experimental status.

I don't know what are the right ways to get out of this situation, but I would appreciate if we could make this feature useful for the interested consumers, it would definitely bring awesome improvements for the community.

I understand where @edolstra and @7c6f434c come from on the risk but I feel like we can be proactive against abuse and be very clear about it. I would even go and say there could be some flag which can disable this feature at runtime with a flag or an environment variable if needed?

Thank you to everyone involved into this!

**Comment by 7c6f434c** [comment]
>I don't know what are the right ways to get out of this situation, but I would appreciate if we could make this feature useful for the interested consumers, it would definitely bring awesome improvements for the community.
>
>I understand where @edolstra and @7c6f434c come from on the risk but I feel like we can be proactive against abuse and be very clear about it. I would even go and say there could be some flag which can disable this feature at runtime with a flag or an environment variable if needed?

For the record, if this is an experimental feature I am in favour of just pretending that the parameter name is not special in case the feature is off, so that the same build happens (just without discarding the references).

The mixed use between systems with and without the feature would actually keep people honest. 

If anything, regression where some expression with no underscore-named settings goes from working to not working on Nix update is not a good thing…

**Comment by Ericson2314** [comment]
Having the same derivation build two different ways depending on experimental feature usage sounds very bad to me. Failing so the success means the same result is much better.

We could add a primop to ask if an experimental feature is enabled (impure mode only until eval cache keys improve), and that will allow ergonomic falling back at a better layer of abstraction.

What is hard about Hydra? If one is running ones own hydra, one can just change the Nix configuration.

**Comment by 7c6f434c** [comment]
>Having the same derivation build two different ways depending on experimental feature usage sounds very bad to me. Failing so the success means the same result is much better.

The _build_ happens exactly in the same way. The only difference is which of the false-positive references are recorded as runtime dependencies.

**Comment by Ericson2314** [comment]
Then s/build/realization process/ or whatever. "Derivation -> set of store objects" is the mapping I want to not depend on which extensions are enabled (modulo failing to produce a key-value pair at all).

**Comment by roberth** [comment]
Another idea is to add a flag to the Nixpkgs `config` so that users who have enabled the experimental feature can opt in explicitly.
That's a bit more machinery than I would like, but it'd work. Something like `nixpkgs.config.experimentalFeatures.unsafeDiscardReferences = true;` on NixOS?

So to summarize, it seems that we have to give up on one of
 - _reproducible_: references will be the same, as long as Nix isn't old
 - _easy to collaborate with non-experimental users_
 - _easy to enable_: only in `nix.conf` and not in Nixpkgs

The last one seems least bad to sacrifice.

**Comment by Ericson2314** [comment]
Yeah I agree experimental features should probably not be changing behavior, more like monotonically increasing behaviors. e.g. when we add something like `__contentAddressed` it is never ignored: we technically break anything that used it before to instead fail if the experimental feature is not enabled.

**Comment by Ericson2314** [comment]
...and in writing that I have a new idea, we could have an opt-in like `__vanillaDrvPath = true` so Nix will set `drvPath` in the right way. This is a very small variation from your idea, but seems like it might make it more "endorsed by Nix" fwiw.

Also on a tangent, instead of periodically stealing `__blah` attrs, we could have some sort of proactive stealing of an entire namespace (sub attribute set with structured attrs? prefix without structured attrs?) and Nix could simply refuse to build a derivaiton with a so-designated magic attribute it doesn't know about. This gives us better forward-compat.

**Comment by 7c6f434c** [comment]
> Then s/build/realization process/ or whatever. "Derivation -> set of store objects" is the mapping I want to not depend on which extensions are enabled (modulo failing to produce a key-value pair at all).

Well, the set of objects in the store here is the same. If the argument for the feature is that the references ignored don't matter, well, then they shouldn't matter in either direction. (And dropping some references when pushing to a cache is already pretty easy).

**Comment by thufschmitt** [comment]
I think it makes sense to just stabilize this per https://nixos.org/manual/nix/stable/contributing/experimental-features.html#lifecycle-of-an-experimental-feature. @NixOS/nix-team, any thoughts on that?

That obviously wouldn't solve all the problems (since everything but the latest Nix versions would still fail on it), so the idea of a nixpkgs config option like @roberth mentioned seems very good. That's fairly similar to what CA derivations have (there's a `nixpkgs.config.contenAddressedByDefault` flag that can be flipped to start using them). That would give us a nice migration path until the required Nix version to evaluate nixpkgs gets recent-enough to have this working and stable.

**Comment by thufschmitt** [comment]
> I think it makes sense to just stabilize this

#8322
